File: Python3/Medium/Accepted/valid-sudoku_8.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        def validRow(board, row):
            seen = set()
            for i in range(9):
                if board[row][i] != '.':
                    if board[row][i] in seen:
                        return False
                    seen.add(board[row][i])
            return True
        
        def validCol(board, col):
            seen = set()
            for i in range(9):
                if board[i][col] != '.':
                    if board[i][col] in seen:
                        return False
                    seen.add(board[i][col])
            return True
        
        def validBox(board, row, col):
            seen = set()
            for i in range(row, row+3):
                for j in range(col, col+3):
                    if board[i][j] != '.':
                        if board[i][j] in seen:
                            return False
                        seen.add(board[i][j])
            return True
        
        for i in range(9):
            if not validRow(board, i) or not validCol(board, i):
                return False
        
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                if not validBox(board, i, j):
                    return False
        
        return True

File: Rust/Medium/Accepted/valid-sudoku_8.rs
impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        

        let mut rows = vec![0; 9];
        let mut cols = vec![0; 9];
        let mut blocks = vec![0; 9];
        for i in 0..9 {
            for j in 0..9 {
                if board[i][j] != '.' {
                    let value = 1 << (board[i][j] as u8 - '1' as u8);
                    let block = i / 3 * 3 + j / 3;
                    if value & rows[i] > 0 || value & cols[j] > 0 || value & blocks[block] > 0 {
                        return false;
                    }
                    rows[i] |= value;
                    cols[j] |= value;
                    blocks[block] |= value;
                }
            }
        }
        true
    }
}